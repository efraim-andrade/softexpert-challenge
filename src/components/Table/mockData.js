export default [
  { symbol: 'SPY', name: 'SPDR S&P 500', price: 314.48 },
  { symbol: 'KMI', name: 'Kinder Morgan Inc.', price: 19.61 },
  { symbol: 'INTC', name: 'Intel Corporation', price: 58.09 },
  { symbol: 'MU', name: 'Micron Technology Inc.', price: 47.59 },
];
