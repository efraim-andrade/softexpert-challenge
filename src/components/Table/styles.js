import styled from 'styled-components';

import { pxToRem } from '~/functions';
import { colors, metrics, fonts } from '~/theme';

export const Container = styled.div``;

export const TableContainer = styled.table`
  width: 100%;

  border-collapse: separate;
  border-spacing: 0 ${metrics.baseSpacing(0.5)};

  > thead {
    background: ${colors.primary};
  }

  > tbody {
    margin-top: ${metrics.baseSpacing()};

    > tr {
      background: ${colors.rowBlueLight};

      &:nth-child(2n) {
        background: ${colors.rowBlueDark};
      }
    }
  }
`;

export const THeadLabel = styled.th`
  height: ${pxToRem(40)};

  ${fonts.size2};
  font-weight: 700;
  color: ${colors.white};
  text-transform: capitalize;

  &:first-child {
    border-top-left-radius: ${metrics.borderRadius()};
  }

  &:last-child {
    border-top-right-radius: ${metrics.borderRadius()};
  }
`;

export const TBodyItem = styled.td`
  ${fonts.size2};
  font-weight: 400;
  text-align: center;
`;

export const CardsWrapper = styled.div``;

export const Card = styled.div`
  margin-bottom: ${metrics.baseSpacing()};
  border-radius: ${metrics.borderRadius()};
  padding: ${metrics.baseSpacing(1.5)} ${metrics.baseSpacing(2)};

  background: ${colors.rowBlueLight};

  &:nth-child(2n) {
    background: ${colors.rowBlueDark};
  }
`;

export const Row = styled.div`
  display: flex;
  align-items: center;
  justify-content: space-between;

  & + div {
    margin-top: ${metrics.baseSpacing()};
  }

  > strong {
    ${fonts.size1};
    font-weight: 700;
    color: ${colors.grey};
    text-transform: capitalize;
  }

  > p {
    ${fonts.size1};
    font-weight: 400;
    text-align: right;
  }
`;
