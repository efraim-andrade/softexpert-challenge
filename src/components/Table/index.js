import React from 'react';

import Search from '../Search';

import { Container, TableContainer, THeadLabel, TBodyItem } from './styles';

export default function Table() {
  function renderDesktop() {
    return (
      <TableContainer>
        <thead>
          <tr>
            <THeadLabel>Symbol</THeadLabel>
            <THeadLabel>Name</THeadLabel>
            <THeadLabel>Price</THeadLabel>
          </tr>
        </thead>

        <tbody>
          <tr>
            <TBodyItem>MMM</TBodyItem>
            <TBodyItem>3M Company</TBodyItem>
            <TBodyItem>$ 127.0</TBodyItem>
          </tr>
          <tr>
            <TBodyItem>MMM</TBodyItem>
            <TBodyItem>3M Company</TBodyItem>
            <TBodyItem>$ 127.0</TBodyItem>
          </tr>
          <tr>
            <TBodyItem>MMM</TBodyItem>
            <TBodyItem>3M Company</TBodyItem>
            <TBodyItem>$ 127.0</TBodyItem>
          </tr>
        </tbody>
      </TableContainer>
    );
  }

  return (
    <Container>
      <Search />

      {renderDesktop()}
    </Container>
  );
}
