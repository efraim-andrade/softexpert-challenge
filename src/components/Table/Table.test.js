import React from 'react';
import { Router } from 'react-router-dom';
import matchMediaPolyfill from 'mq-polyfill';
import { createMemoryHistory } from 'history';
import { render, cleanup } from '@testing-library/react';

import Table from './index';
import data from './mockData';

describe('Components - Table', () => {
  const originalError = console.error;

  beforeAll(() => {
    matchMediaPolyfill(window);

    window.resizeTo = function resizeTo(width, height) {
      Object.assign(this, {
        innerWidth: width,
        innerHeight: height,
        outerWidth: width,
        outerHeight: height,
      }).dispatchEvent(new this.Event('resize'));
    };

    console.error = (...args) => {
      if (/Warning.*not wrapped in act/.test(args[0])) {
        return;
      }
      originalError.call(console, ...args);
    };
  });

  afterEach(cleanup);

  afterAll(() => {
    console.error = originalError;
  });

  const history = createMemoryHistory();

  it('Should render correctly', () => {
    const { asFragment } = render(
      <Router history={history}>
        <Table tableData={data} allTableData={data} setTableData={() => {}} />
      </Router>
    );

    expect(asFragment()).toMatchSnapshot();
  });

  it('Should change the table layout when window width is less than 720px', () => {
    window.resizeTo(720, 1000);

    const { getByTestId } = render(
      <Router history={history}>
        <Table tableData={data} allTableData={data} setTableData={() => {}} />
      </Router>
    );

    const layoutMobile = getByTestId('layout-mobile');

    expect(layoutMobile).toBeDefined();
  });

  it("Should be appears an message when don't have data", () => {
    const { getByText } = render(
      <Router history={history}>
        <Table tableData={[]} allTableData={[]} setTableData={() => {}} />
      </Router>
    );

    const theMessage = getByText('Nothing was found...');

    expect(theMessage).toBeDefined();
  });
});
