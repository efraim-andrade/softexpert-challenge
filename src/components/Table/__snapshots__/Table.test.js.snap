// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Components - Table Should render correctly 1`] = `
<DocumentFragment>
  <div
    class="sc-htpNat fXqbef"
  >
    <div
      class="sc-bwzfXH VVOgA"
    >
      <button
        type="button"
      >
        <svg
          aria-hidden="true"
          class="sc-bdVaJa iJCmgd"
          color="#0091ED9B"
          fill="none"
          focusable="false"
          height="24"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          viewBox="0 0 24 24"
          width="24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <circle
            cx="11"
            cy="11"
            r="8"
          />
          <line
            x1="21"
            x2="16.65"
            y1="21"
            y2="16.65"
          />
        </svg>
      </button>
      <input
        placeholder="Search something..."
      />
    </div>
    <table
      class="sc-bxivhb lnGjjs"
    >
      <thead>
        <tr>
          <th
            class="sc-ifAKCX eLAIso"
          >
            symbol
          </th>
          <th
            class="sc-ifAKCX eLAIso"
          >
            name
          </th>
          <th
            class="sc-ifAKCX eLAIso"
          >
            price
          </th>
        </tr>
      </thead>
      <tbody>
        <tr>
          <td
            class="sc-EHOje jwqwWK"
          >
            SPY
          </td>
          <td
            class="sc-EHOje jwqwWK"
          >
            SPDR S&P 500
          </td>
          <td
            class="sc-EHOje jwqwWK"
          >
            $ 314.48
          </td>
        </tr>
        <tr>
          <td
            class="sc-EHOje jwqwWK"
          >
            KMI
          </td>
          <td
            class="sc-EHOje jwqwWK"
          >
            Kinder Morgan Inc.
          </td>
          <td
            class="sc-EHOje jwqwWK"
          >
            $ 19.61
          </td>
        </tr>
        <tr>
          <td
            class="sc-EHOje jwqwWK"
          >
            INTC
          </td>
          <td
            class="sc-EHOje jwqwWK"
          >
            Intel Corporation
          </td>
          <td
            class="sc-EHOje jwqwWK"
          >
            $ 58.09
          </td>
        </tr>
        <tr>
          <td
            class="sc-EHOje jwqwWK"
          >
            MU
          </td>
          <td
            class="sc-EHOje jwqwWK"
          >
            Micron Technology Inc.
          </td>
          <td
            class="sc-EHOje jwqwWK"
          >
            $ 47.59
          </td>
        </tr>
      </tbody>
    </table>
  </div>
</DocumentFragment>
`;
